storage/s3_test.go

```diff
@@ -398,3 +404,150 @@ func TestS3Retry(t *testing.T) {
 		})
 	}
 }
+
+// Credit to aws-sdk-go
+func val(i interface{}, s string) interface{} {
+	v, err := awsutil.ValuesAtPath(i, s)
+	if err != nil || len(v) == 0 {
+		return nil
+	}
+	if _, ok := v[0].(io.Reader); ok {
+		return v[0]
+	}
+
+	if rv := reflect.ValueOf(v[0]); rv.Kind() == reflect.Ptr {
+		return rv.Elem().Interface()
+	}
+
+	return v[0]
+}
+
+func TestS3AclFlagOnCopy(t *testing.T) {
+	testcases := []struct {
+		name string
+		acl  string
+
+		expectedAcl string
+	}{
+		{
+			name: "no acl flag",
+		},
+		{
+			name: "acl flag without value, flag should be ignored",
+		},
+		{
+			name:        "acl flag with a value",
+			acl:         "bucket-owner-full-control",
+			expectedAcl: "bucket-owner-full-control",
+		},
+	}
+	defaultReqErr := fmt.Errorf("request was canceled by the user")
```

Comment:
Let's remove `defaultReqErr` and update mocked response as :

```go
r.HTTPResponse = &http.Response{
	StatusCode: http.StatusOK,
	Body:       ioutil.NopCloser(strings.NewReader("")),
}
```

Returning  an error for each `mockS3.Copy` call is a bit confusing.

Same for `put` tests.